### PR TITLE
build: update scaffolding for creating unary math iterator packages

### PR DESCRIPTION
**Pull request by gunjjoshi**
## Description

> What is the purpose of this pull request?

This pull request:

-   adds `runner.js` in [`_tools/scaffold/math-iter-unary/scripts`](https://github.com/stdlib-js/stdlib/tree/develop/lib/node_modules/%40stdlib/_tools/scaffold/math-iter-unary/scripts) to generate packages in [`math/iter/special`](https://github.com/stdlib-js/stdlib/tree/develop/lib/node_modules/%40stdlib/math/iter/special).

## Related Issues

> Does this pull request have any related issues?

This pull request:

-   resolves a part of #1147.

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

### CODE REVIEW

**Comment by gunjjoshi** [comment]
While running this, I get the following error:

Screenshot 2024-09-21 at 00 16 34

**Comment by gunjjoshi** [comment]
I thought to update [`scaffold.sh`](https://github.com/stdlib-js/stdlib/blob/develop/lib/node_modules/%40stdlib/_tools/scaffold/math-iter-unary/scripts/scaffold.sh) once we have a basic working script.

Along with that, are we also thinking to update our scaffolding structure (https://github.com/stdlib-js/stdlib/tree/develop/lib/node_modules/%40stdlib/_tools/scaffold/math-iter-unary/data), as our meta data does not contain some of the fields used in that.

File: lib/node_modules/@stdlib/_tools/scaffold/math-iter-unary/scripts/runner.js [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+		if ( existsSync( p ) ) {
+			if ( CREATE_ONLY ) {
+				log( 'Package already exists. Skipping @%s...', 'stdlib/math/iter/special/' + o.alias );
+				continue;
+			}
+			log( 'Updating package: @%s...', 'stdlib/math/iter/special/' + o.alias );
+		} else {
+			log( 'Creating package: @%s...', 'stdlib/math/iter/special/' + o.alias );
+		}
+		names = objectKeys( o );
```
Comment (kgryte) [inline lib/node_modules/@stdlib/_tools/scaffold/math-iter-unary/scripts/runner.js]: @gunjjoshi You need to be more selective in what you extract from the JSON file. You need to map the JSON value to a corresponding environment variable in the scaffold script. You cannot just copy-paste the `runner.js` from the `random-array-unary` scaffold.

File: lib/node_modules/@stdlib/_tools/scaffold/math-iter-unary/scripts/runner.js [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+		for ( j = 0; j < names.length; j++ ) {
+			k = names[ j ];
+			v = o[ k ];
+			str = uppercase( k );
+			str += '=';
+			str += '\'' + v + '\'';
+			envs.push( str );
+		}
+		cmd = envs.join( ' ' ) + ' . ' + SCAFFOLD_SCRIPT;
+		shell( cmd );
```
Comment (kgryte) [inline lib/node_modules/@stdlib/_tools/scaffold/math-iter-unary/scripts/runner.js:275]: During development, you can print the expected command by commenting out this line and adding a `console.log( cmd )`.

**Comment by gunjjoshi** [comment]
What are we using this for: https://github.com/stdlib-js/stdlib/blob/develop/lib/node_modules/%40stdlib/_tools/scaffold/math-iter-unary/scripts/data.json#L3 ?
With this included, I get the error: 
```
/bin/sh: =math/base@v1.0: No such file or directory
```

When I remove this from `data.json`, it works, and the packages are being made at the required path.

File: lib/node_modules/@stdlib/_tools/scaffold/math-iter-unary/scripts/runner.js [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+		names = objectKeys( o );
+		envs = [];
+		for ( j = 0; j < names.length; j++ ) {
+			k = names[ j ];
+			v = o[ k ];
+			str = uppercase( k );
+			str += '=';
+			str += '\'' + v + '\'';
+			envs.push( str );
+		}
```
Comment (kgryte) [inline lib/node_modules/@stdlib/_tools/scaffold/math-iter-unary/scripts/runner.js]: @gunjjoshi The reason why you are getting the errors is that you are still passing everything to the `scaffold.sh` script. You need to explicitly map fields in the `data.json` to environment variables in the `scaffold.sh` file. We shouldn't expect these to be 1:1. Meaning, you actually need to perform data transformations.

File: lib/node_modules/@stdlib/_tools/scaffold/math-iter-unary/scripts/scaffold.sh [hunk 4]
```diff
@@ -104,7 +104,7 @@ camelcase_to_snakecase() {
 }
 
 # Convert the alias to snakecase:
-alias_snakecase=$(camelcase_to_snakecase "${alias:4}")
+alias_snakecase=$(camelcase_to_snakecase "${alias}")
```
Comment (kgryte) [inline lib/node_modules/@stdlib/_tools/scaffold/math-iter-unary/scripts/scaffold.sh]: It's not clear why you'd need to change this or remove the unary function alias logic below.

File: lib/node_modules/@stdlib/_tools/scaffold/math-iter-unary/scripts/runner.js (hunk 3 above)
Comment (kgryte) [reply lib/node_modules/@stdlib/_tools/scaffold/math-iter-unary/scripts/runner.js]: Furthermore, I wouldn't expect that we'd need everything in the `data.json` file. I would expect that we pick and choose the appropriate data fields.

Comment (gunjjoshi) [reply lib/node_modules/@stdlib/_tools/scaffold/math-iter-unary/scripts/runner.js]: Oh, got it. Thanks!

File: lib/node_modules/@stdlib/_tools/scaffold/math-iter-unary/scripts/scaffold.sh (hunk 4 above)
Comment (gunjjoshi) [reply lib/node_modules/@stdlib/_tools/scaffold/math-iter-unary/scripts/scaffold.sh]: Understood this now, I've reverted this change.

**Comment by gunjjoshi** [comment]
Made a few changes, the structure is generated correctly now.

File: lib/node_modules/@stdlib/_tools/scaffold/math-iter-unary/scripts/scaffold.sh [hunk 5]
```diff
@@ -147,7 +147,7 @@ unary_pkg_path="stdlib/math/base/special"
 unary_pkg="${unary_pkg_path}/${unary_pkg_alias}"
 
 # Define the REPL text main export description (note: 4 space indent and wrap at 80 characters):
-repl_text_main_desc=$(echo -n "${main_desc}" | "${wrap}")
+repl_text_main_desc=$(echo "${main_desc}" | "${wrap}")
```
Comment (kgryte) [inline lib/node_modules/@stdlib/_tools/scaffold/math-iter-unary/scripts/scaffold.sh]: @gunjjoshi Why was the `-n` flag removed here? Having this suppresses appending a newline character to printed output.

File: lib/node_modules/@stdlib/_tools/scaffold/math-iter-unary/scripts/runner.js [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+		} else {
+			log( 'Creating package: @%s...', 'stdlib/math/iter/special/' + o.alias );
+		}
+		names = [ 'base_alias', 'alias', 'pkg_desc', 'desc', 'example_values', 'rand', 'keywords' ];
+		envs = [];
+		for ( j = 0; j < names.length; j++ ) {
+			k = names[ j ];
+			v = o[ k ];
+			if ( k === 'example_values' ) {
+				envs.push( 'VALUES_LEN_2=\''+o.parameters[0].example_values.join(', ')+'\'' );
```
Comment (kgryte) [inline lib/node_modules/@stdlib/_tools/scaffold/math-iter-unary/scripts/runner.js]:
```suggestion
				envs.push( 'VALUES_LEN_2=\''+o.parameters[0].example_values.join(',')+'\'' );
```

File: lib/node_modules/@stdlib/_tools/scaffold/math-iter-unary/scripts/scaffold.sh [hunk 7]
```diff
@@ -37,23 +37,23 @@
 ## USER-DEFINED VARIABLES ##
 
 # Define the main export alias:
-alias=${ALIAS:-'iterTest'}
+alias='iter'$(echo ${ALIAS:-'Test'} | awk '{print toupper(substr($0, 1, 1)) tolower(substr($0, 2))}')
```
Comment (kgryte) [inline lib/node_modules/@stdlib/_tools/scaffold/math-iter-unary/scripts/scaffold.sh]: I think my preference would be to move the alias logic to JavaScript. While it is common for us to do `iterXxxxx`, it is not always the case. So I'd revert to

```
${ALIAS:-'iterTest'}
```

and then in the runner.js file do

```
envs.push( 'ALIAS=iter'+capitalize( v ) );
```

or something similar.

File: lib/node_modules/@stdlib/_tools/scaffold/math-iter-unary/scripts/scaffold.sh (hunk 5 above)
Comment (gunjjoshi) [reply lib/node_modules/@stdlib/_tools/scaffold/math-iter-unary/scripts/scaffold.sh]: Keeping `-n` leads to an addition of `-n` in the output, something as follows:

```
-n Returns an iterator which iteratively computes the cube root.
```

File: lib/node_modules/@stdlib/_tools/scaffold/math-iter-unary/scripts/scaffold.sh (hunk 7 above)
Comment (gunjjoshi) [reply lib/node_modules/@stdlib/_tools/scaffold/math-iter-unary/scripts/scaffold.sh]: Done, thanks!

File: lib/node_modules/@stdlib/_tools/scaffold/math-iter-unary/scripts/scaffold.sh (hunk 5 above)
Comment (kgryte) [reply lib/node_modules/@stdlib/_tools/scaffold/math-iter-unary/scripts/scaffold.sh]: I see. The issue is the `-n` is not portable across all Linux shells.

Comment (kgryte) [inline lib/node_modules/@stdlib/_tools/scaffold/math-iter-unary/scripts/scaffold.sh]:
```suggestion
repl_text_main_desc=$(printf "%s" "${main_desc}" | "${wrap}")
```

We can replace with `printf`, I believe.

File: lib/node_modules/@stdlib/_tools/scaffold/math-iter-unary/scripts/scaffold.sh [hunk 8]
```diff
@@ -41,19 +41,19 @@ alias=${ALIAS:-'iterTest'}
 
 # Define the package description:
 # shellcheck disable=SC2016
-pkg_desc=${PKG_DESC:-'Create an iterator which computes `gamma(x+1) - 1` for each iterated value.'}
+pkg_desc="Create an iterator which ${DESC:-'computes \`gamma(x+1) - 1\`'} for each iterated value."
```
Comment (kgryte) [inline lib/node_modules/@stdlib/_tools/scaffold/math-iter-unary/scripts/scaffold.sh]:
```suggestion
pkg_desc="Create an iterator which ${DESC:-'TODO'} for each iterated value."
```

Let's add a placeholder.

File: lib/node_modules/@stdlib/_tools/scaffold/math-iter-unary/scripts/scaffold.sh [hunk 9]
```diff
@@ -41,19 +41,19 @@ alias=${ALIAS:-'iterTest'}
...
 
 # Define the module description:
 # shellcheck disable=SC2016
-module_desc=${MODULE_DESC:-'Create an iterator which iteratively computes `gamma(x+1) - 1`.'}
+module_desc="Create an iterator which iteratively ${DESC:-'computes \`gamma(x+1) - 1\`'}."
```
Comment (kgryte) [inline lib/node_modules/@stdlib/_tools/scaffold/math-iter-unary/scripts/scaffold.sh]:
```suggestion
module_desc="Create an iterator which iteratively ${DESC:-'TODO'}."
```

File: lib/node_modules/@stdlib/_tools/scaffold/math-iter-unary/scripts/scaffold.sh [hunk 10]
```diff
@@ -41,19 +41,19 @@ alias=${ALIAS:-'iterTest'}
...
 
 # Define main export description:
 # shellcheck disable=SC2016
-main_desc=${MAIN_DESC:-'Returns an iterator which iteratively computes `gamma(x+1) - 1` without cancellation errors for small `x`.'}
+main_desc="Returns an iterator which iteratively ${DESC:-'computes \`gamma(x+1) - 1\`'}."
```
Comment (kgryte) [inline lib/node_modules/@stdlib/_tools/scaffold/math-iter-unary/scripts/scaffold.sh]:
```suggestion
main_desc="Returns an iterator which iteratively ${DESC:-'TODO'}."
```

File: lib/node_modules/@stdlib/_tools/scaffold/math-iter-unary/scripts/scaffold.sh [hunk 11]
```diff
@@ -41,19 +41,19 @@ alias=${ALIAS:-'iterTest'}
...
 
 # Define the test description:
 # shellcheck disable=SC2016
-test_desc=${TEST_DESC:-'computes `gamma(x+1) - 1` for each iterated value'}
+test_desc=${DESC:-'computes \`gamma(x+1) - 1\`'}'for each iterated value'
```
Comment (kgryte) [inline lib/node_modules/@stdlib/_tools/scaffold/math-iter-unary/scripts/scaffold.sh]:
```suggestion
test_desc=${DESC:-'TODO'}' for each iterated value'
```

**Review by kgryte — APPROVED** [review]
Okay. I think this is now ready for merge. I made a number of changes, @gunjjoshi. In preparation for other scaffolding tools (which will get more complex!), probably worthwhile studying the changes and reverse engineering why I made the ones that I did.

Some things won't be obvious (e.g., replacing backticks in `require` paths in `index.d.ts` files => historical artifact due to prior tooling limitation), while others rely a bit more on understanding that there will be exceptions we'll need to account for (e.g., relying on `DESC` to provide a default description will work for many cases, but there will be other packages in which we cannot do this and need to be more particular about specifying `PKG_DESC`, `MODULE_DESC`, etc).

**Comment by gunjjoshi** [comment]
Thanks for making these changes, @kgryte. I'll go through them once so that we follow the latest conventions in future packages.

**Pull request merged**